### PR TITLE
Fail message for data::Load in test files

### PR DESCRIPTION
**Pull request by heisenbuug**
I hope it passes all the check, but anyways most of the work is done, we can make manual changes if required any.

### CODE REVIEW

**Comment by Abilityguy** [comment]
Hi @heisenbuug, thanks for taking the time to do this. 
Taking a quick look at the pipelines, it seems there are some errors due to your changes.

For future changes, I would say it would be a good to batch changes in multiple PRs, since it would make it easier for the maintainers to review.

**Comment by heisenbuug** [comment]
> Hi @heisenbuug, thanks for taking the time to do this.
> Taking a quick look at the pipelines, it seems there are some errors due to your changes.
> 
> For future changes, I would say it would be a good to batch changes in multiple PRs, since it would make it easier for the maintainers to review.

Sorry for mixing it all, I will make sure to divide it into multiple PRs, my bad

File: src/mlpack/tests/io_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/io_test.cpp:417]:
```suggestion
    FAIL("Cannot load test dataset test.csv!"); 
```

File: src/mlpack/tests/kfn_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/kfn_test.cpp:470]:
```suggestion
    FAIL("Cannot load test dataset test_data_3_1000.csv!"); 
```

File: src/mlpack/tests/krann_search_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/krann_search_test.cpp:39]:
```suggestion
    FAIL("Cannot load test dataset test_r_3_900.csv!"); 
```

**Comment by heisenbuug** [comment]
> Hi @heisenbuug, thanks for taking the time to do this.
> Not sure if the fails are related to this PR.
> 
> For future changes, I would say it would be a good to batch changes in multiple PRs, since it would make it easier for the maintainers to review.

@Abilityguy So, I think it's done for now? 
Also, I later read the discussion about maintaining `Cannot open dataset...`, so I will make changes accordingly

File: src/mlpack/tests/lsh_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/lsh_test.cpp:121]:
```suggestion
    FAIL("Cannot load test dataset iris_train.csv! ");
```
Some issue with the script here?
It says ```rainSe```.

In this test case, they are loading the ```iris_train.csv``` and dataset ```iris_test.csv```. 
It would be better if the message included the name of the dataset.

**Review by Abilityguy — CHANGES_REQUESTED** [review]
Hi @heisenbuug, I took a look at the files and have suggested some changes. 

It would be great if we change all the messages to ```Cannot load test dataset ... !``` since it maintains consistency with the older files.

**Comment by heisenbuug** [comment]
> Hi @heisenbuug, I took a look at the files and have suggested some changes.
I made all the changes @Abilityguy, I hope everything is done now
> Also, it seems like the fails are not due to the changes in this PR.
That's great then. Thank you for the help!!!

**Comment by Abilityguy** [comment]
Hey @heisenbuug , your changes include a lot of addition and deletion whitespaces. 
If you take a look at the file changes [here](https://github.com/mlpack/mlpack/pull/2717/files), you will notice a lot of unnecessary changes that the script has made.

Maybe you could take a look at those?

File: src/mlpack/methods/ann/layer/adaptive_max_pooling.hpp
Comment (Abilityguy) [inline src/mlpack/methods/ann/layer/adaptive_max_pooling.hpp:53]:
```suggestion
  AdaptiveMaxPooling(const std::tuple<size_t, size_t>& outputShape);

```
Even though the whitespaces here were unnecessary, the newline is required.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:33]:
```suggestion
                 MatType& trainData,
```
Indentation needs to be fixed here.

**Comment by heisenbuug** [comment]
> Hey @heisenbuug , your changes include a lot of addition and deletion whitespaces.
> If you take a look at the file changes [here](https://github.com/mlpack/mlpack/pull/2717/files), you will notice a lot of unnecessary changes that the script has made.
> 
> Maybe you could take a look at those?

Done with the changes.
`src/mlpack/tests/feedforward_network_test.cpp` this file somehow is causing many issues, don't know why I think I messed up the commit somewhere, but everything is resolved now.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:33]:
```suggestion
                 MatType& trainData,
```

Hey @heisenbuug , indentation issue still persists here and at other places.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:61]:
```suggestion
                       MatType& trainData,
```
Indentation issue here.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:84]:
```suggestion
                       MatType& trainData,
```
Indentation issue here.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:252]:
```suggestion
  opt.Initialize(model.Parameters().n_rows, model.Parameters().n_cols);
```

Indentation issue here.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:255]:
```suggestion
      model.Parameters().n_rows, model.Parameters().n_cols);
```
Indentation issue here.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:431]: Indentation issue in the model comment here.

File: src/mlpack/tests/load_save_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/load_save_test.cpp:1462]: Would it be possible to add the dataset name here?
testFiles is a vector of strings so the ```FAIL()``` message could include the dataset name.
```testFiles[i]``` should give the name of the dataset that failed.

**Review by Abilityguy — COMMENTED** [review]
Hi @heisenbuug , pointed out a few more style fixes with the code.
Maybe you could manually fix the files with spacing issues since the script seems to adding them.

File: src/mlpack/tests/load_save_test.cpp
Comment (heisenbuug) [reply src/mlpack/tests/load_save_test.cpp:1462]: We can put the value of testFiles[i] in a string and then pass the string to the fail message, does that sound good?

Comment (Abilityguy) [reply src/mlpack/tests/load_save_test.cpp:1462]: Yup. That sounds good. 

> And sorry to trouble you with all the style checks.

Not an issue. 👍🏼

Comment (Abilityguy) [reply src/mlpack/tests/load_save_test.cpp]: Hi. I think I was not clear enough when I pointed out the indentation issues.
I can only mark a line so I pointed the start of the line.
But the entire block of code is actually indented wrong. 
You can get a better idea when you see the code diff [here](https://github.com/mlpack/mlpack/pull/2717/files).

Comment (heisenbuug) [reply src/mlpack/tests/load_save_test.cpp]: Implemented that failed message as discussed...

Comment (heisenbuug) [reply src/mlpack/tests/load_save_test.cpp]: Working on it...

> Hi. I think I was not clear enough when I pointed out the indentation issues.
> I can only mark a line so I pointed the start of the line.
> But the entire block of code is actually indented wrong.
> You can get a better idea when you see the code diff [here](https://github.com/mlpack/mlpack/pull/2717/files).

Comment (heisenbuug) [reply src/mlpack/tests/load_save_test.cpp]: I think it's done now.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:326]: Indentation issue from lines 326-346

File: src/mlpack/tests/load_save_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/load_save_test.cpp:88]: Indentation issue in lines 88 and 89.

File: src/mlpack/tests/load_save_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/load_save_test.cpp:1417]: Lines 1417-1429 need to indented another two spaces to maintain the indentation of the block.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:718]: Let's add back this removed newline to maintain consistency.

**Review by Abilityguy — CHANGES_REQUESTED** [review]
Good job @heisenbuug! 
I believe we are close to finishing this.
Just pointed out a few minor changes and I think we are done with the style fixes.

**Review by Abilityguy — APPROVED** [review]
The changes look good to me. The failing checks seem to be unrelated. 
Let's see what the maintainers have to say.

**Review by rcurtin — COMMENTED** [review]
Thanks @heisenbuug and @Abilityguy for going over this in such detail!  I think that this handles all of the remaining files, so we can close #2715 after this is merged.  Let me know if I'm mistaken on that.  All the style and changes look good but I just have one comment.

Now, I do think that in cases where we have `data::Load(filename, matrix, true)`, that `true` indicates that a fatal error should be issued if loading fails.  That should throw an exception in that case.  For instance, here's what happens with the NBC tests:

```
$ bin/mlpack_test "[NBCTest]"
mlpack version: mlpack git-075d6c4be
armadillo version: 9.900.1 (Nocturnal Misbehaviour)
Filters: [NBCTest]
[FATAL] Cannot open file 'trainSet.csv'. 


~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
mlpack_test is a Catch v2.13.3 host application.
Run with -? for options

-------------------------------------------------------------------------------
NaiveBayesClassifierTest
-------------------------------------------------------------------------------
/home/ryan/src/mlpack/src/mlpack/tests/nbc_test.cpp:19
...............................................................................

/home/ryan/src/mlpack/src/mlpack/tests/nbc_test.cpp:19: FAILED:
due to unexpected exception with message:
  fatal error; see Log::Fatal output

[FATAL] Cannot open file 'trainSet.csv'. 

...
```

(that's without these changes)

Anyway, I think it's fine to make these changes and use a `FAIL()` call instead of depending on the exception-throwing support of `data::Load()`.  The output is probably nicer.  But if we want the output to be from `FAIL()` whenever loading fails, we should find any calls where `true` is set in the last argument of `data::Load()` and just remove that third argument.  (So, e.g., `data::Load("thyroid_train.csv", trainData, true)` --> `data::Load("thyroid_train.csv", trainData)`.)

Thanks again for taking the time to work on this! :+1:

**Comment by Abilityguy** [comment]
You are right @rcurtin. Merging this PR will close the issue #2715. 
Thanks for pointing out the ```fatal``` argument issue. 

Going through the files, there is another case 
```data::Load("rann_test_qr_ranks.csv", qrRanks, true, false);``` where the fourth argument allows us to transpose the loaded matrix. 

For such cases, I think the appropriate change is
```data::Load(x, y, true, false);``` --> ```data::Load(x, y, false, false);```?

Correct me if I am wrong here.

**Comment by rcurtin** [comment]
@Abilityguy right, what you suggested is the correct change to make for the four-argument case. :+1:

File: src/mlpack/tests/feedforward_network_2_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_2_test.cpp:68]:
```suggestion
  if (!data::Load("thyroid_train.csv", trainData)) 
```
Third argument not required.

File: src/mlpack/tests/feedforward_network_2_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_2_test.cpp:81]:
```suggestion
  if (!data::Load("thyroid_test.csv", testData)) 
```
Third argument not required.

File: src/mlpack/tests/feedforward_network_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/feedforward_network_test.cpp:110]: Third argument not required.

File: src/mlpack/tests/io_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/io_test.cpp:841]:
```suggestion
  if (!data::Load("test.csv", dataset2, false, false)) 
```
Third argument should be changed to 'false'.

File: src/mlpack/tests/krann_search_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/krann_search_test.cpp:38]:
```suggestion
  if (!data::Load("rann_test_r_3_900.csv", refData)) 
```

File: src/mlpack/tests/krann_search_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/krann_search_test.cpp:46]:
```suggestion
  if (!data::Load("rann_test_qr_ranks.csv", qrRanks, false, false)) 
```

File: src/mlpack/tests/ksinit_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/ksinit_test.cpp:233]:
```suggestion
  if (!data::Load("iris.csv", dataset)) 
```
Third argument not required.

File: src/mlpack/tests/load_save_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/load_save_test.cpp:1467]:
```suggestion
    if (!data::Load(testFiles[i], one, false, false))
```

File: src/mlpack/tests/load_save_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/load_save_test.cpp:2037]:
```suggestion
  if (!data::Load("test.csv", dataset, info, false, true)) 
```

File: src/mlpack/tests/lrsdp_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/lrsdp_test.cpp:98]:
```suggestion
  if (!data::Load("johnson8-4-4.csv", edges)) 
```
Third argument not required.

File: src/mlpack/tests/lsh_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/lsh_test.cpp:120]:
```suggestion
  if (!data::Load(trainSet, rdata)) 
```
Third argument not required.

File: src/mlpack/tests/matrix_completion_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/matrix_completion_test.cpp:40]:
```suggestion
  if (!data::Load("completion_X.csv", Xorig, false, false)) 
```

File: src/mlpack/tests/nbc_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/nbc_test.cpp:31]:
```suggestion
  if (!data::Load(trainResultFilename, trainRes)) 
```
Third argument not required.

File: src/mlpack/tests/nystroem_method_test.cpp
Comment (Abilityguy) [inline src/mlpack/tests/nystroem_method_test.cpp:149]:
```suggestion
  if (!data::Load("german.csv", dataset)) 
```
Third argument not required.

**Review by Abilityguy — CHANGES_REQUESTED** [review]
Hey @heisenbuug , I have pointed out a few examples in files where changes need to be made. Could you take a look and make the changes in all the required files?

**Comment by heisenbuug** [comment]
> Hey @heisenbuug , I have pointed out a few examples in files where changes need to be made. Could you take a look and make the changes in all the required files?

Hey @Abilityguy, I will work on the required changes and will make a PR soon.

**Comment by Abilityguy** [comment]
> > Hey @heisenbuug , I have pointed out a few examples in files where changes need to be made. Could you take a look and make the changes in all the required files?
> 
> Hey @Abilityguy, I will work on the required changes and will make a PR soon.

I don't think you need to make a new PR. The changes can be made in this PR itself.

**Comment by heisenbuug** [comment]
Heyy @Abilityguy, I messed up something, and now its showing me this
`heisenbuug wants to merge 49 commits into mlpack:master from unknown repository`
What should I do? 
Shall I make a new PR?
Sorry for the trouble...

**Comment by Abilityguy** [comment]
Not an issue @heisenbuug .
You can make a new PR and cherry pick the commits from this PR to get to the state this PR is in. 
This [gist](https://gist.github.com/ozh/cbce675ba35e0d336cff) provides quick info on how to cherry pick.
Do get back if you need any help.

**Comment by Abilityguy** [comment]
Hi @heisenbuug , if you are still finding it difficult to make new commits, what you can do is just commit the suggestions I have made above in this thread using the GUI.

**Pull request closed without merging**
